src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1547,3 +2145,348 @@ MultiRouterPlannableQuery(Query *query, MultiExecutorType taskExecutorType,
+			TargetEntry *oldSubqueryTle = list_nth(subquery->targetList,
+												   oldInsertVar->varattno - 1);
+
+			newSubqueryTargetEntry = copyObject(oldSubqueryTle);
+
+			newSubqueryTargetEntry->resno = resno;
+			newSubqueryTargetlist = lappend(newSubqueryTargetlist,
+											newSubqueryTargetEntry);
+
+			updatedSubqueryEntryCount++;
+		}
+		else
+		{
+			newSubqueryTargetEntry = makeTargetEntry(oldInsertTargetEntry->expr,
+													 resno,
+													 oldInsertTargetEntry->resname,
+													 oldInsertTargetEntry->resjunk);
+			newSubqueryTargetlist = lappend(newSubqueryTargetlist,
+											newSubqueryTargetEntry);
+		}
+
+		newInsertVar = makeVar(insertTableId, originalAttrNo,
+							   exprType((Node *) newSubqueryTargetEntry->expr),
+							   exprTypmod((Node *) newSubqueryTargetEntry->expr),
+							   exprCollation((Node *) newSubqueryTargetEntry->expr),
+							   0);
+		newInsertTargetEntry = makeTargetEntry((Expr *) newInsertVar, originalAttrNo,
+											   oldInsertTargetEntry->resname,
+											   oldInsertTargetEntry->resjunk);
+
+		newInsertTargetlist = lappend(newInsertTargetlist, newInsertTargetEntry);
+		resno++;
+	}
+
+	/*
+	 * if there are any remaining target list entries (i.e., GROUP BY column not on the
+	 * target list of subquery), update the remaining resnos.
+	 */
+	subqueryTargetLength = list_length(subquery->targetList);
+	if (subqueryTargetLength != updatedSubqueryEntryCount)
+	{
+		int targetEntryIndex = updatedSubqueryEntryCount;
+
+		for (; targetEntryIndex < subqueryTargetLength; ++targetEntryIndex)
+		{
+			TargetEntry *oldSubqueryTle = list_nth(subquery->targetList,
+												   targetEntryIndex);
+			TargetEntry *newSubqueryTargetEntry = copyObject(oldSubqueryTle);
+
+			Assert(newSubqueryTargetEntry->resjunk == true);
+
+			newSubqueryTargetEntry->resno = resno;
+			newSubqueryTargetlist = lappend(newSubqueryTargetlist,
+											newSubqueryTargetEntry);
+
+			resno++;
+		}
+	}
+
+	originalQuery->targetList = newInsertTargetlist;
+	subquery->targetList = newSubqueryTargetlist;
+
+	return NULL;
+}
+
+
+/*
+ * InsertSelectQuery returns true when the input query
+ * is INSERT INTO ... SELECT kind of query.
+ */
+bool
+InsertSelectQuery(Query *query)
+{
+	CmdType commandType = query->commandType;
+	List *rangeTableList = query->rtable;
+	RangeTblEntry *subqueryRte = NULL;
+	Query *subquery = NULL;
+
+	if (commandType != CMD_INSERT)
+	{
+		return false;
+	}
+
+	rangeTableList = query->rtable;
+	if (list_length(rangeTableList) < 2)
+	{
+		return false;
+	}
+
+	subqueryRte = lsecond(query->rtable);
+	subquery = subqueryRte->subquery;
+	if (subquery == NULL)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * Copy a RelationRestrictionContext. Note that several subfields are copied
+ * shallowly, for lack of copyObject support.
+ */
+static RelationRestrictionContext *
+copyRelationRestrictionContext(RelationRestrictionContext *oldContext)
+{
```

Comment:
Hmm it looks like we also use `restriction->->relOptInfo->joininfo` in router planner and  it is copyable. So, I added that.

I'm also thinking to update the name of the function such that the reader could understand that we only copy the parts of the context that we use in the code later (either read or write). Does that sound reasonable?